[DTensor] add layer norm backward support

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116437
* #116428
* #115689
* __->__ #115683

**Summary**
This PR adds DTensor implementation for ATen op `native_layer_norm_backward`.


**Test Plan**
pytest test/distributed/_tensor/test_math_ops.py -s -k layer_norm
pytest test/distributed/_tensor/test_dtensor_ops.py

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

XilunWu committed:
[DTensor] add layer norm backward support

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52093619).

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -350,6 +351,164 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+            # and normalized input (x_hat) should have the same sharding
+            # placements, and grad_out's sharding is determined by the
+            # pointwise result of x_hat and weight/bias.
+            d_input_strategy.strategies[-1].input_specs = op_args_target_specs
```
why we only associate input specs to the last strategy?

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -350,6 +351,164 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+            d_weight_strategy.strategies[-1].input_specs = op_args_target_specs
+
+        if output_mask[2]:
+            d_bias_strategy.strategies[-1].input_specs = op_args_target_specs
```
hmmm thinking about this more, looks like the input specs are exactly the same for each OpStrategy output?

This make me think that TupleStrategy might be a overkill and headable to maintain for cases where we have "multiple different outputs that have different sharding strategies"

Maybe we should piggyback to OpStrategy, and simply make PlacementStrategy.output_spec be a `Union[DTensorSpec, Tuple[Optional[DTensorSpec], ...]]`

wanchaol reviewed (commented):
The op implementation looks legit, however after reading the code I feel using TupleStrategy brings quite a bit of complexity there. Maybe we should simply make PlacementStrategy.output_spec to support multiple outputs... Let's discuss more on this tmr.

yoyoyocmu commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -350,6 +351,164 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
 
 
+@register_op_strategy(
+    [aten.native_layer_norm_backward.default],
```
Thanks for adding the backward support. I think we also need to remove the existing `aten.native_layer_norm_backward.default` from experimental_op.py.

XilunWu replied:
@yoyoyocmu you're right. Regarding correctness, the rule in this PR has higher priority than the experimental one (the experimental rule is shadowed). I will remove the experimental op in #115689

XilunWu committed:
Update on "[DTensor] add layer norm backward support"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093619](https://our.internmc.facebook.com/intern/diff/D52093619)

[ghstack-poisoned]

XilunWu commented:
put this PR back to draft because the weight/bias gradient is not correctly handled -- not keeping them as Partial

XilunWu committed:
Update on "[DTensor] add layer norm backward support"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093619](https://our.internmc.facebook.com/intern/diff/D52093619)

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -350,6 +351,164 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+        # of replicating the input tensor.
+        if output_mask[0]:
+            grad_out_src_spec = grad_out_strategy.strategies[idx].output_spec
+            grad_out_target_spec = DTensorSpec(
```
will address this with lint fix.

XilunWu replied:
after a second thought, I prefer to keeping the arg naming consistent with `native_layer_norm_backward` as in `torch/_decomp/decompositions.py` for people to better understand. WDYT?

wanchaol replied:
hmmm shouldn't it call `d_weight_spec` instead? I don't get quite what the `grad_out_target_spec` indicates here

XilunWu commented:
may need to fix extra lint errors since my local lint has issue

XilunWu committed:
Update on "[DTensor] add layer norm backward support"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093619](https://our.internmc.facebook.com/intern/diff/D52093619)

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] add layer norm backward support"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093619](https://our.internmc.facebook.com/intern/diff/D52093619)

[ghstack-poisoned]

wanchaol commented on test/distributed/_tensor/test_math_ops.py:
```diff
@@ -200,7 +209,100 @@ def _replicate_fn(name, module, device_mesh):
+            self.assertEqual(y_local, y_dist.full_tensor())
+
+            # backward step
+            comm_mode = CommDebugMode()
```
I think you can reuse the `comm_mode` initialize earlier instead of creating a new one, the collective counts will be cleared for a new context

wanchaol commented on test/distributed/_tensor/test_math_ops.py:
```diff
@@ -200,7 +209,100 @@ def _replicate_fn(name, module, device_mesh):
+                    layer_norm_dist.weight.grad.full_tensor(),
+                )
+                self.assertEqual(
+                    layer_norm_local.bias.grad,
```
can we also assert that the dist.weight/grad are partial tensors?

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -354,6 +354,178 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+        output_specs_list: List[Optional[DTensorSpec]] = []
+        op_args_target_specs = []
+        redistribute_costs = []
+        input_src_spec = input_strategy.strategies[idx].output_spec
```
if you are already accessing input_strategy.strategies[idx], can't we assign a variable in the for loop instead of the current `for idx, _`?

XilunWu replied:
good catch!

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -354,6 +354,178 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+            redistribute_costs.append(
+                generate_redistribute_costs(weight_strategy, weight_target_spec)
+            )
+            # weight_grad_spec = weight_target_spec  # TODO
```
this comment is invalid now?

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -354,6 +354,178 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+            )
+            output_specs_list.append(grad_out_target_spec)
+            out_tuple_strategy.strategies.append(
+                # these 3 lists will be filled in later
```
It's less intuitive to first create a PlacementStrategy with all empty lists and then fill the PlacementStrategy. could we create PlacementStrategy fill the out_tuple_strategy at the very end of the for loop when we fill the output_specs_list and op_args_target_specs?

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -354,6 +354,178 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+        assert isinstance(mean_target_spec, DTensorSpec)
+        op_args_target_specs.append(mean_target_spec)
+        redistribute_costs.append(
+            generate_redistribute_costs(mean_strategy, mean_target_spec)
```
given that we don't change the mean_target_spec at all, the redistribute_cost should simply be 
[0.0] * len(mean_strategy.strategies)?

XilunWu replied:
I see. This means we do not pass down the `redistribute_cost` along the computation chain? For example, ((a + b) * c), the redistribute cost from `+` won't be passed to `*`.

wanchaol replied:
yes we don't need to pass down redistribute cost along the chain, even with the current change I believe the redistribute_cost generated by this function call are all 0s

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -354,6 +354,178 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+        assert isinstance(rstd_target_spec, DTensorSpec)
+        op_args_target_specs.append(rstd_target_spec)
+        redistribute_costs.append(
+            generate_redistribute_costs(rstd_strategy, rstd_target_spec)
```
ditto for rstd strategy, we can just make the costs be 0.0

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -354,6 +354,178 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+            )
+            op_args_target_specs.append(weight_target_spec)
+            redistribute_costs.append(
+                generate_redistribute_costs(weight_strategy, weight_target_spec)
```
same, weight spec redistribute cost seems 0 too

XilunWu replied:
Because it's always "Replicate"?

wanchaol replied:
because the weight_target_spec come from weight_strategy (one of the weight strategy), which means they should be idential

XilunWu replied:
oh that's right. I keep confusing `weight` and `d_weight`...

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -354,6 +354,178 @@ def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+            )
+            op_args_target_specs.append(bias_target_spec)
+            redistribute_costs.append(
+                generate_redistribute_costs(bias_strategy, bias_target_spec)
```
ditto: bias redistribute cost seems 0

wanchaol commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -71,7 +71,9 @@ def register_op_strategy(
         self, op_schema: OpSchema
-    ) -> Union[None, TensorMeta, List[TensorMeta], Tuple[TensorMeta, ...]]:
+    ) -> Union[
+        None, TensorMeta, List[Optional[TensorMeta]], Tuple[Optional[TensorMeta], ...]
```
For this type annotation, we can unify to `Sequence[Optional[TensorMeta]]`?

wanchaol commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -117,7 +121,10 @@ def _wrap_output_spec_tensor_meta(
+            None,
+            TensorMeta,
+            List[Optional[TensorMeta]],
+            Tuple[Optional[TensorMeta], ...],
```
`Sequence[Optional[TensorMeta]]`

wanchaol commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -279,12 +286,19 @@ def _get_output_spec_from_op_strategy(
+                        assert isinstance(output_strategy.output_spec, DTensorSpec)
+                        out_spec_list.append(output_strategy.output_spec)
+                    else:
+                        # for None output in tuple, its output spec should also be None
```
we don't need to change TupleStrategy handling anymore?

wanchaol reviewed (commented):
This looks much better, please see inlined comments

XilunWu committed:
Update on "[DTensor] add layer norm backward support"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093619](https://our.internmc.facebook.com/intern/diff/D52093619)

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] add layer norm backward support"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093619](https://our.internmc.facebook.com/intern/diff/D52093619)

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] add layer norm backward support"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093619](https://our.internmc.facebook.com/intern/diff/D52093619)

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] add layer norm backward support"


**Summary**
This PR adds DTensor implementation for ATen op `native_layer_norm_backward`.


**Test Plan**
pytest test/distributed/_tensor/test_math_ops.py -s -k layer_norm
pytest test/distributed/_tensor/test_dtensor_ops.py

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093619](https://our.internmc.facebook.com/intern/diff/D52093619)

[ghstack-poisoned]

wanchaol commented on test/distributed/_tensor/test_math_ops.py:
```diff
@@ -200,7 +210,112 @@ def _replicate_fn(name, module, device_mesh):
+                # and bias should be _Partial
+                dim_map = x_dist._spec.dim_map
+                outer_dims = range(norm_idx)
+                needs_reduction = any(dim_map[d] >= 0 for d in outer_dims)
```
Can we just use `assert isinstance(grad.placements[0], Partial)`. 

The `dim_map` property would not include partial sum informations so the check here seems wrong?

wanchaol commented on test/distributed/_tensor/test_math_ops.py:
```diff
@@ -200,7 +210,112 @@ def _replicate_fn(name, module, device_mesh):
+            if elementwise_affine:
+                # if input is sharded on any outer dimension, the gradient of weight
+                # and bias should be _Partial
+                dim_map = x_dist._spec.dim_map
```
we probably shouldn't use `dim_map` as it's not well defined for partial sum

wanchaol reviewed (commented):
Some comments inlined

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52093619).

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52093619).

wanchaol reviewed (approved):
synced offline, lgtm

XilunWu committed:
Update on "[DTensor] add layer norm backward support"


**Summary**
This PR adds DTensor implementation for ATen op `native_layer_norm_backward`.


**Test Plan**
pytest test/distributed/_tensor/test_math_ops.py -s -k layer_norm
pytest test/distributed/_tensor/test_dtensor_ops.py

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225


[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] add layer norm backward support"


**Summary**
This PR adds DTensor implementation for ATen op `native_layer_norm_backward`.


**Test Plan**
pytest test/distributed/_tensor/test_math_ops.py -s -k layer_norm
pytest test/distributed/_tensor/test_dtensor_ops.py

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225


[ghstack-poisoned]
